irohad/main/application.hpp

```diff
@@ -57,68 +64,116 @@ class Irohad {
-      pb_query_response_factory,
-      std::shared_ptr<iroha::torii::QueryProcessor> query_processor);
-
-  std::shared_ptr<iroha::torii::QueryProcessor> createQueryProcessor(
-      std::unique_ptr<iroha::model::QueryProcessingFactory> qpf,
-      std::shared_ptr<iroha::validation::StatelessValidator>
-      stateless_validator);
-
-  std::shared_ptr<iroha::torii::TransactionProcessor>
-  createTransactionProcessor(
-      std::shared_ptr<iroha::network::PeerCommunicationService> pcs,
-      std::shared_ptr<iroha::validation::StatelessValidator> validator);
-
-  std::shared_ptr<iroha::validation::StatelessValidator>
-  createStatelessValidator(
-      std::shared_ptr<iroha::model::ModelCryptoProvider> crypto_provider);
-
-  std::unique_ptr<iroha::model::QueryProcessingFactory>
-  createQueryProcessingFactory(
-      std::shared_ptr<iroha::ametsuchi::WsvQuery> wsvQuery,
-      std::shared_ptr<iroha::ametsuchi::BlockQuery> blockQuery);
+ protected:
+// ------------------------| component initialization |-------------------------
+
+  virtual void initStorage();
+
+  virtual void initLoop();
+
+  virtual void initProtoFactories();
+
+  virtual void initCryptoProvider();
+
+  virtual void initHashProvider();
+
+  virtual void initValidators();
+
+  virtual void initPeerQuery();
+
+  virtual void initPeerOrderer();
+
+  virtual void initPeerAddress();
+
+  virtual void initOrderingGate();
+
+  virtual void initSimulator();
+
+  virtual void initBlockLoader();
 
+  virtual void initConsensusGate();
+
+  virtual void initSynchronizer();
+
+  virtual void initPeerCommunicationService();
+
+  virtual void initTransactionCommandService();
+
+  virtual void initQueryService();
+
+  // constructor dependencies
   std::string block_store_dir_;
   std::string redis_host_;
   size_t redis_port_;
   std::string pg_conn_;
   size_t torii_port_;
+
+// ---------------------------| internal decencies |----------------------------
+
+  // loop
   std::shared_ptr<uvw::Loop> loop;
 
-  std::unique_ptr<::torii::CommandService> command_service;
-  std::unique_ptr<::torii::QueryService> query_service;
+  // converter factories
+  std::shared_ptr<iroha::model::converters::PbTransactionFactory> pb_tx_factory;
+  std::shared_ptr<iroha::model::converters::PbQueryFactory> pb_query_factory;
+  std::shared_ptr<iroha::model::converters::PbQueryResponseFactory> pb_query_response_factory;
+
+  // crypto provider
+  std::shared_ptr<iroha::model::ModelCryptoProvider> crypto_verifier;
+
+  // hash provider
+  std::shared_ptr<iroha::model::HashProviderImpl> hash_provider;
+
+  // validators
+  std::shared_ptr<iroha::validation::StatelessValidator> stateless_validator;
+  std::shared_ptr<iroha::validation::StatefulValidatorImpl> stateful_validator;
+  std::shared_ptr<iroha::validation::ChainValidatorImpl> chain_validator;
+
+  // peer query
+  std::shared_ptr<iroha::ametsuchi::PeerQueryWsv> wsv;
+
+  // peer orderer
+  std::shared_ptr<iroha::consensus::yac::PeerOrdererImpl> orderer;
```

Comment:
Maybe use interface here instead of implementation? Same for some other fields of this class.